src/blob/index/bptree/core.rs

```diff
@@ -376,46 +383,51 @@ where
         Ok(())
     }
 
-    async fn validate_header(&self, buf: &mut Vec<u8>, blob_size: u64) -> Result<()> {
+    async fn validate_header(&self, buf: BytesMut, blob_size: u64) -> Result<BytesMut> {
         self.validate(blob_size)?;
-        if !Self::hash_valid(&self.header, buf)? {
+        if let Some(buf) = Self::hash_valid(&self.header, buf)? {
+            Ok(buf)
+        } else {
             let param = ValidationErrorKind::IndexChecksum;
-            return Err(Error::validation(param, "header hash mismatch").into());
+            Err(Error::validation(param, "header hash mismatch").into())
         }
-        Ok(())
     }
 
-    fn hash_valid(header: &IndexHeader, buf: &mut Vec<u8>) -> Result<bool> {
+    fn hash_valid(header: &IndexHeader, mut buf: BytesMut) -> Result<Option<BytesMut>> {
         let hash = header.hash.clone();
+        println!("Before {:?}", buf.len());
         let mut header = header.clone();
         header.hash = vec![0; ring::digest::SHA256.output_len];
         header.set_written(false);
-        serialize_into(buf.as_mut_slice(), &header)?;
+        serialize_into(&mut buf[..], &header)?;
         let new_hash = get_hash(&buf);
-        Ok(hash == new_hash)
+        println!("After {:?}", buf.len());
+        if hash == new_hash {
+            return Ok(Some(buf));
+        } else {
+            return Ok(None);
+        }
     }
 
     async fn read_index_header(file: &File) -> Result<IndexHeader> {
         let header_size = IndexHeader::serialized_size_default()? as usize;
-        let mut buf = vec![0; header_size];
-        file.read_at(&mut buf, 0).await?;
+        let buf = file.read_at_allocate(header_size, 0).await?;
         IndexHeader::from_raw(&buf).map_err(Into::into)
     }
 
-    async fn read_root(file: &File, root_offset: u64) -> Result<[u8; BLOCK_SIZE]> {
-        let mut buf = [0; BLOCK_SIZE];
+    async fn read_root(file: &File, root_offset: u64) -> Result<BytesMut> {
         let buf_size = std::cmp::min((file.size() - root_offset) as usize, BLOCK_SIZE);
-        file.read_at(&mut buf[..buf_size], root_offset).await?;
+        let mut buf = file.read_at_allocate(buf_size, root_offset).await?;
+        buf.resize(BLOCK_SIZE, 0);
```

Comment:
If `buf_size < BLOCK_SIZE`, then here the allocation happens to resize the buffer. I suggest to allocate `buf` initially with `BLOCK_SIZE` length, then truncate before  `read_at_allocate` and then resize back. In this way the internal capacity not changes and no reallocation happens